### PR TITLE
Yul optimizer test: Separate test parsing and optimisation

### PR DESCRIPTION
**Pull request by bshastry**
This PR separates yul parsing from optimisation in the test framework.

Required by #10603

### CODE REVIEW

File: test/libyul/YulOptimizerTestCommon.cpp [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+		ForLoopInitRewriter::run(*m_context, *m_ast);
+	}
+	else if (m_optimizerStep == "commonSubexpressionEliminator")
+	{
+		disambiguate();
+		if (m_fuzzerMode)
+		{
+			ForLoopInitRewriter::run(*m_context, *m_ast);
+			FunctionHoister::run(*m_context, *m_ast);
+		}
```
Comment (bshastry) [inline test/libyul/YulOptimizerTestCommon.cpp]: The fuzzer may synthesize for loops with init blocks and/or nested function definitions. Hence this `if` block

File: test/libyul/YulOptimizerTestCommon.cpp [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+		return false;
+
+	return true;
+}
+
+shared_ptr<Block> YulOptimizerTester::run()
+{
+	bool success = runStep();
+	if (!success)
+		return nullptr;
```
Comment (hrkrshnn) [inline test/libyul/YulOptimizerTestCommon.cpp]:
```suggestion
	return nullptr;
```
Too much indentation?

File: test/libyul/YulOptimizerTest.h [hunk 3]
```diff
@@ -56,22 +56,15 @@ class YulOptimizerTest: public solidity::frontend::test::EVMVersionRestrictedTes
...
-	void disambiguate();
-	void updateContext();
-
 	static void printErrors(std::ostream& _stream, langutil::ErrorList const& _errors);
 
 	std::string m_optimizerStep;
 
 	Dialect const* m_dialect = nullptr;
-	std::set<YulString> m_reservedIdentifiers;
-	std::unique_ptr<NameDispenser> m_nameDispenser;
```
Comment (hrkrshnn) [inline test/libyul/YulOptimizerTest.h:73]: The header in line 24 can be removed

File: test/libyul/YulOptimizerTest.h [hunk 4]
```diff
@@ -56,22 +56,15 @@ class YulOptimizerTest: public solidity::frontend::test::EVMVersionRestrictedTes
...
-	std::unique_ptr<OptimiserStepContext> m_context;
```
Comment (hrkrshnn) [inline test/libyul/YulOptimizerTest.h:74]: Similarly for this.

Comment (hrkrshnn) [reply test/libyul/YulOptimizerTest.h:74]: Actually can't lines 31-43 be removed as well?

File: test/libyul/YulOptimizerTestCommon.cpp (hunk 1 above)
Comment (hrkrshnn) [reply test/libyul/YulOptimizerTestCommon.cpp]: Actually, I'm not sure if we really need a `m_fuzzerMode`. I think it's okay to have these run during the regular tests, unless we were also testing to see what happens when it is executed without, say, `ForLoopInitRewriter`. Can you try setting `m_fuzzerMode` true for the regular tests and try to see if `isoltest` produces any differences?

File: test/libyul/YulOptimizerTestCommon.cpp (hunk 2 above)
Comment (bshastry) [reply test/libyul/YulOptimizerTestCommon.cpp]: Hmm... I just checked locally that it's a single tab. Do you think it's additional spaces?

**Comment by bshastry** [comment]
@hrkrshnn Regarding your comment (https://github.com/ethereum/solidity/pull/10699#discussion_r549660291), the additional steps added in this PR for `fuzzer_mode` did produce different optimised snippets (at least they didn't exactly match, I didn't take a closer look) and hence the flag. Would you like to see what this changes are? There are tens of test cases that are different though.

My main intention here is to keep the status quo while permitting fuzzing to take place in parallel. We could make them uniform in a follow-up PR where the test differences can be reviewed in greater detail?

File: test/libyul/YulOptimizerTestCommon.cpp (hunk 2 above)
Comment (hrkrshnn) [reply test/libyul/YulOptimizerTestCommon.cpp]: Ah nevermind, I thought it was too long and copied it incorrectly while checking.

File: test/libyul/YulOptimizerTest.h (hunk 4 above)
Comment (bshastry) [reply test/libyul/YulOptimizerTest.h:74]: Cleaned up redundant includes and forward declarations.

**Comment by bshastry** [comment]
> at least they didn't exactly match,

Here are the tests that need to be updated if the so called `fuzzer_mode` is unconditionally enabled (https://gist.github.com/bshastry/78c8746806c95e3cdb441329a6f340a3)

**Comment by bshastry** [comment]
@hrkrshnn pointed out that `fuzzer_mode` is not necessary and that the newly introduced changes may be run unconditionally even in the yul test framework. He also pointed out that @ekpyron was also in favour of at least running FLIR where necessary. The other step that is necessary as a pre-req is FunctionHoister for cases when fuzzer synthesizes a function definition inside another function definition.

If that is the case, I could (in this or a follow up PR) get rid of `fuzzer_mode` and update tests accordingly.

**Comment by leonardoalt** [comment]
@bshastry are you waiting for @hrkrshnn 's input?

**Comment by bshastry** [comment]
@leonardoalt Yes, if @hrkrshnn approves and there is no veto against the approach taken by this PR from @ekpyron and @chriseth I'd believe this one is good to be merged.

File: test/libyul/YulOptimizerTestCommon.cpp (hunk 1 above)
Comment (bshastry) [reply test/libyul/YulOptimizerTestCommon.cpp]: We discussed this on gitter: The answer to

> Can you try setting m_fuzzerMode true for the regular tests and try to see if isoltest produces any differences?

Yes

**Comment by hrkrshnn** [comment]
This looks good. But I think it's nice to merge both the refactoring along with just enabling the `fuzzer_mode` part by default, i.e., without the `if`s at the same time.

**Comment by bshastry** [comment]
> This looks good. But I think it's nice to merge both the refactoring along with just enabling the `fuzzer_mode` part by default, i.e., without the `if`s at the same time.

Do you mean just remove the `fuzzer_mode` and have the test framework execute the same sequence of tests for unit tests as well and update tests?

**Comment by hrkrshnn** [comment]
> Do you mean just remove the fuzzer_mode and have the test framework execute the same sequence of tests for unit tests as well and update tests?

Yes. It's fine to do this in a separate PR, and we can merge them together.

**Comment by bshastry** [comment]
> > Do you mean just remove the fuzzer_mode and have the test framework execute the same sequence of tests for unit tests as well and update tests?
> 
> Yes. It's fine to do this in a separate PR, and we can merge them together.

Created https://github.com/ethereum/solidity/pull/10719

File: test/libyul/YulOptimizerTestCommon.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+namespace solidity::yul
+{
+	struct AsmAnalysisInfo;
+	struct Object;
+	struct Dialect;
+}
+
+namespace solidity::yul::test
+{
+class YulOptimizerTester
```
Comment (hrkrshnn) [inline test/libyul/YulOptimizerTestCommon.h]: Does it make sense to rename it to `YulOptimizerTestCommon`? There is already `YulOptimizerTest`

File: test/libyul/YulOptimizerTest.cpp [hunk 6]
```diff
@@ -128,312 +73,12 @@ TestCase::TestResult YulOptimizerTest::run(ostream& _stream, string const& _line
...
-				ASTWalker::operator()(_identifier);
-			}
-			YulString m_currentFunction = YulString{};
-		};
-		FakeUnreachableGenerator fakeUnreachableGenerator;
-		fakeUnreachableGenerator(*obj.code);
-		StackLimitEvader::run(*m_context, obj, fakeUnreachableGenerator.fakeUnreachables);
-	}
-	else
+	if (!success)
```
Comment (hrkrshnn) [inline test/libyul/YulOptimizerTest.cpp]:
```suggestion
	if (!tester.runSetup)
```
and removing line 79?

File: test/libyul/YulOptimizerTestCommon.cpp [hunk 7]
```diff
@@ -0,0 +1,465 @@
...
```
Comment (hrkrshnn) [inline test/libyul/YulOptimizerTestCommon.cpp]: Similarly, here.

**Review by hrkrshnn — COMMENTED** [review]
Minor comments

File: test/libyul/YulOptimizerTestCommon.h (hunk 5 above)
Comment (bshastry) [reply test/libyul/YulOptimizerTestCommon.h]: Done

File: test/libyul/YulOptimizerTestCommon.cpp [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+	else
+		return false;
+
+	return true;
+}
+
+shared_ptr<Block> YulOptimizerTestCommon::run()
+{
+	bool success = runStep();
+	if (!success)
```
Comment (hrkrshnn) [inline test/libyul/YulOptimizerTestCommon.cpp]:
```suggestion
	if (!runStep())
```

Comment (bshastry) [reply test/libyul/YulOptimizerTestCommon.cpp]: Done!

**Comment by leonardoalt** [comment]
Please rebase

**Comment by chriseth** [comment]
Approach looks good in general, but I did not take a deep look.

**Comment by bshastry** [comment]
Rebased

**Review by hrkrshnn — APPROVED** [review]

**Pull request merged**